### ISSUE TITLE
8.8 Playback over WAVE Baseline Splice Constraints 

### ISSUE THREAD

**Issue by jpiesing** [cta-wave/dpctf-tests#22]
Two requirements for test 8.8 are extremely unclear to me.

> Two CMAF Switching Sets are available for playback following the properties in clause5.3.3.5.

What are the characteristics of these? Do we have the right mezzanine content?

In the March 2nd meeting, it was suggested we might want both big buck bunny and tears of steel mezzanine content but how can we reach a decision?

> The playback has the following parameters
> • min_buffer_duration: Expresses the minimum buffer that the Source Buffer maintains in the playback.
> • playout[i]: Provides the triple (Switching Set, CMAF track number, Fragment number) for every playout position i=1,…,N that is be played out.

Who is responsible for providing the value of the playout array?

**Comment by louaybassbouss** [comment]
@dsilhavy FYI

**Comment by jpiesing** [comment]
@nicholas-fr @rbouqueau this may require addtional test content which in turn requires additional mezzanine content. For example, it might need both tears of steel and big buck bunny.

**Comment by jpiesing** [comment]
I've re-read the WAVE content format spec and device playback spec sections relating to baseline splice constraints. 

My first impression is that a CMAF switching set with 30s of some other content encoded using the same parameters as the switching set used for test 8.5 would be sufficient.

The test for 8.8 would play the switching set for test 8.5, then the new switching set, then back to the switching set from test 8.5.

Do people agree with this?

8.8 will be run separately for video and audio but it would be good for the content to be re-usable for 9.4 which tests the combination. Test 9.4 would probably be the same as the DASH-IF issues https://github.com/Dash-Industry-Forum/Test-Content/issues/1 and following.

Focusing on 9.4 for now & inspired by the DASH-IF issue, it might be best to use 30s of Big Buck Bunny.

In order to meet the requirements for 9.4, it seems to me that the video and audio CMAF fragment lengths would need to be somehow aligned. In DVB, it was mentioned that multiples of 960ms may work well for video and (some) audio codecs.

For DASH-IF test cases 2, https://github.com/Dash-Industry-Forum/Test-Content/issues/2 and https://github.com/Dash-Industry-Forum/Test-Content/issues/3, we would want to avoid perfect alignment between video and audio fragment lengths at the ends of **all** video fragments. 

For DASH-IF test cases 1 and 2, the additional test content would need the video and audio fragment lengths aligned.
Although the additional test content has to be a switching set, it's not obvious why it would need more than one bitrate.

**Comment by louaybassbouss** [comment]
Thanks @jpiesing for your investigation I will discuss this tomorrow with @dsilhavy especially in regards to the DASH-IF Test Content issues and will provide feedback.

**Comment by andyburras** [comment]
Does this mean the frame number sequence will run 1->X->1->Y->1->Z ?
Presumably the test config file will need to pass the expected duration of each splice "section".

**Comment by jpiesing** [comment]
> Does this mean the frame number sequence will run 1->X->1->Y->1->Z ?

Either that or 1 -> X-1 -> Y -> X -> Z?

> Presumably the test config file will need to pass the expected duration of each splice "section".

I have no idea.

**Comment by dsilhavy** [comment]
Some comments from our side, as I discussed this with @louaybassbouss 

 I find the description in DPCTF 8.8 a bit confusing. Speaking in DASH terminology, it sounds like we are testing a switch of AdaptationSets with similar media types within a single period. So for instance:
  - Audio: Switch between audio language
  - Video: Switch between viewing angle / camera

If this interpretation is correct I suggest to use the same mezzanine root content for this use case and for video maybe flip it by 90 or 180 degrees. That way, there is a clear difference between the SwitchingSets/AdaptationSets but no "real" transition to a totally different content.

However, the result of this test case is important for 9.4 as the same logic applies.

As Jon mentioned: DPCTF 9.4 seems to map to an ad-insertion test case as we are switching between CMAF presentations. From my understanding CMAF presentations can be interpreted as DASH periods or single period DASH MPDs combined via MPD chaining. In this case it corresponds to https://github.com/Dash-Industry-Forum/Test-Content/issues/1 and the following issues. For this, I suggest to use two different types of mezzanine content like BBB and ToS.

Some general comments:

- As described by Jon, media segment alignment can be achieved with segment durations of 1.92 seconds for 25fps and 48khz content for example. No promotion intended, but we gave some more examples in this blog post: https://websites.fraunhofer.de/video-dev/why-and-how-to-align-media-segments-for-abr-streaming/
- It would be great if DASH-IF can reuse the WAVE content for the test cases defined here: https://github.com/Dash-Industry-Forum/Test-Content/issues. For this reason, I think we need:
  - Three different mezzanine files (main content, ad, ad slate). For example BBB, ToS, and a spinning WAVE/DASH-IF logo
  - For each mezzanine file, at least two different segment durations to create the overlaps described in https://github.com/Dash-Industry-Forum/Test-Content/issues/2 and https://github.com/Dash-Industry-Forum/Test-Content/issues/3

**Comment by jpiesing** [comment]
> Some comments from our side, as I discussed this with @louaybassbouss
> 
> I find the description in DPCTF 8.8 a bit confusing.

I'm glad I'm not the only one :)

> Speaking in DASH terminology, it sounds like we are testing a switch of AdaptationSets with similar media types within a single period. So for instance:
> 
>     * Audio: Switch between audio language
> 
>     * Video: Switch between viewing angle / camera
> 

That's not how I read DPCTF 5.3.3.5 and CSTF 7.2.2. I think these are the equivalent of DASH period-connected but I certainly admit I could be wrong. 

> 
> If this interpretation is correct I suggest to use the same mezzanine root content for this use case and for video maybe flip it by 90 or 180 degrees. That way, there is a clear difference between the SwitchingSets/AdaptationSets but no "real" transition to a totally different content.
> 
> However, the result of this test case is important for 9.4 as the same logic applies.
> 

I ended up seeing 8.8 as a subset of 9.4 and hence trying to work out what is wanted for 9.4 and then working backwards to 8.8.

> As Jon mentioned: DPCTF 9.4 seems to map to an ad-insertion test case as we are switching between CMAF presentations. From my understanding CMAF presentations can be interpreted as DASH periods or single period DASH MPDs combined via MPD chaining. In this case it corresponds to [Dash-Industry-Forum/Test-Content#1](https://github.com/Dash-Industry-Forum/Test-Content/issues/1) and the following issues. For this, I suggest to use two different types of mezzanine content like BBB and ToS.
> 

I think this is a multi-period MPD with the periods being period-connected.
WAVE already has mezzanine content for ToS. It is generated by a script so running a version of that script on BBB would not be a lot of work.

> Some general comments:
> 
>     * As described by Jon, media segment alignment can be achieved with segment durations of 1.92 seconds for 25fps and 48khz content for example. No promotion intended, but we gave some more examples in this blog post: https://websites.fraunhofer.de/video-dev/why-and-how-to-align-media-segments-for-abr-streaming/
> 
>     * It would be great if DASH-IF can reuse the WAVE content for the test cases defined here: https://github.com/Dash-Industry-Forum/Test-Content/issues. For this reason, I think we need:
>       

I agree. It would be good if HbbTV could use this as well. 

>       * Three different mezzanine files (main content, ad, ad slate). For example BBB, ToS, and a spinning WAVE/DASH-IF logo

Yes. Perhaps DVB and HbbTV logos too :)

>       * For each mezzanine file, at least two different segment durations to create the overlaps described in [Dash-Industry-Forum/Test-Content#2](https://github.com/Dash-Industry-Forum/Test-Content/issues/2) and [Dash-Industry-Forum/Test-Content#3](https://github.com/Dash-Industry-Forum/Test-Content/issues/3)

Do you have any suggestions? I'm currently working on a document on how to generate WAVE test content from the mezzanines.

**Comment by jpiesing** [comment]
@dsilhavy Presumably the same original streams can cover https://github.com/Dash-Industry-Forum/Test-Content/issues/1 but with the first period duration being a multiple of 1.92?

**Comment by louaybassbouss** [comment]
@rbouqueau do we have now multi-period for this content (as described in  https://github.com/cta-wave/dpctf-tests/issues/22#issuecomment-795091964)?

**Comment by rbouqueau** [comment]
Daniel just sent me his feedback on the streams generated for https://github.com/Dash-Industry-Forum/Test-Content/issues/1 this morning. Will process and let you know.

**Comment by louaybassbouss** [comment]
> > @rbouqueau do we have now multi-period for this content (as described in [#22 (comment)](https://github.com/cta-wave/dpctf-tests/issues/22#issuecomment-795091964))?
> 
> @louaybassbouss @rbouqueau I wasn't expecting multi-period content with a single MPD but two separate MPDs, one for the main content & one for the insertion.

@jpiesing the test runner accepts only one MPD for video and one for audio. This is why we decided to use multi-period for this test (see discussion in https://github.com/cta-wave/Test-Content-Generation/issues/25#issuecomment-801016251). The multi-period is generated  from the two single MPDs. The media segments from both single-period content will be reused.

**Comment by rbouqueau** [comment]
We need to make adjustments as described [here](https://github.com/cta-wave/dpctf-tests/issues/22#issuecomment-795091964) but ok. Could we please open an issue on the Test Generation repo when it is confirmed this is needed?

**Comment by jpiesing** [comment]
I have re-reviewed section 7 and annex A of the CSTF spec concerning splice conditioning
My conclusion is that what is described there is indeed the same as DASH-IF issue 1 - https://github.com/Dash-Industry-Forum/Test-Content/issues/1 .
A single MPD with 3 Periods would indeed address what is described in the CSTF spec. 
This isn't what the DPCTF spec says though so we need to change the spec.
The comments above refer to 3 pieces of content;
- Main content could be the existing WAVE mezzanine content
- @nicholas-fr has created mezzanine content based on BBB - see https://github.com/cta-wave/mezzanine/issues/18 and streams beginning ad_bbb in https://dash.akamaized.net/WAVE/Mezzanine/new/ .
- I don't quite understand the ad slate.

**Comment by nicholas-fr** [comment]
The mezzanine content for this was moved here:
https://dash.akamaized.net/WAVE/Mezzanine/new/2021-03-31/
Feedback on the labels to use, duration, and other aspects are welcome. How the ad slate would be used is also unclear to me. Some clarification would be useful before we create any additional mezzanine content for that.

**Comment by dsilhavy** [comment]
Please ignore if this was not your question: 

Ad slates are used to fill small gaps between the announced duration of an ad pod and the duration of the ad period returned by the ad server. For example:

- Duration of the ad pod is 30 seconds
- The ad server returns an ad with a duration of 28 seconds
- The remaining 2 seconds are filled with an ad slate. After the slate the player returns to the main program.

If possible I suggest to use a spinning logo for the slates. For instances the Wave logo, the DASH-IF logo or just a caption saying "ad-slate". It would be nice though if the logo is moving as this makes it easier to identify stalls.

**Comment by jpiesing** [comment]
@dsilhavy @louaybassbouss  What Period structure would you propose for 8.8? Essentially the WAVE splicing conditions are DASH-IF test case 1 - without any overlaps. Could you write a sufficiently detailed description for @rbouqueau to script it?

**Comment by andyburras** [comment]
@dsilhavy So just to clarify... the ad slate is not required for the CTA WAVE DPCTF testing because we are not testing players, just the device's underlying implementations (so there is no difference below the MSE API between a splice from main->ad and from ad->slate).
However, the ad slate would be very useful to allow DASH-IF testing to make use of the same CTA WAVE test content (where players may be being tested)?

**Comment by jpiesing** [comment]
@rbouqueau @louaybassbouss I'm trying to work out how this divides between your two activities.
@louaybassbouss Please can you be extremely precise about what you're waiting for? My guess is that it's two pieces of test content 1) according to table 5 & table 6 of CTA 5001, 2) with the same content options, 3) with video & audio segment durations aligned to that the second content can be inserted into the middle of the first with zero overlap at either end.

**Comment by dsilhavy** [comment]
- Let's leave the ad slate out of the scope here. As you mentioned, from an MSE perspective there is no difference between transitioning from main to ad and ad to ad slate. The difference here lies in the small periods for ad slates. These need to be handled correctly by a DASH player. Consequently, out of scope for this discussion.
- The content created by @rbouqueau for https://github.com/Dash-Industry-Forum/Test-Content/issues/1 , specifically https://dash.akamaized.net/dashif/ad-insertion-testcase1/batch5/counter/a/ad-insertion-testcase1.mpd can be reused here. However, I suggest to clarify/change:
  1. Clarify: Check if the mezzanine content for the main periods (1 and 3) is ok. 
  2. Change: Replace the mezzanine content for the ad period with the new BBB mezzanine content.
  3. Clarify: Define whether period 1 and 3 are continuous. By continuous I mean if period 3 resumes where period 1 left off. Right now the duration of the ad period is added to the @presentationTimeOffset of period 3. You can see that when playing in dash.js: https://reference.dashif.org/dash.js/nightly/samples/dash-if-reference-player/index.html?mpd=https://dash.akamaized.net/dashif/ad-insertion-testcase1/batch5/counter/a/ad-insertion-testcase1.mpd.  The timestamp of period 3 starts around 19 seconds. I would argue that this is the standard behavior for live ad insertion but not for VoD ad insertion. For VoD ad insertion period 3 would resume at 9.6 seconds (where period 1 stopped).
- The duration of the segments is ok, they are aligned (1.92 seconds each). Consequently we have no overlap.

To summarize what we need:
- Agreement on the 3 points above 
- A concrete url in this thread that should be used for the tests. Then we can remove the waiting label.

From my understanding the code for the test is already in place. @louaybassbouss please correct me here.

**Comment by jpiesing** [comment]
>     1. Clarify: Check if the mezzanine content for the main periods (1 and 3) is ok.

The observation framework relies on the QR codes (etc) in the WAVE mezzanine so we would need to use that mezzanine for the main periods.

>     3\. Clarify: Define whether period 1 and 3 are continuous. By continuous I mean if period 3 resumes where period 1 left off. Right now the duration of the ad period is added to the @presentationTimeOffset of period 3. You can see that when playing in dash.js: https://reference.dashif.org/dash.js/nightly/samples/dash-if-reference-player/index.html?mpd=https://dash.akamaized.net/dashif/ad-insertion-testcase1/batch5/counter/a/ad-insertion-testcase1.mpd.  The timestamp of period 3 starts around 19 seconds. I would argue that this is the standard behavior for live ad insertion but not for VoD ad insertion. For VoD ad insertion period 3 would resume at 9.6 seconds (where period 1 stopped).

Which is more likely to cause a problem with an MSE implementation? Adding the duration of the ad or not adding it?

**Comment by dsilhavy** [comment]
> Which is more likely to cause a problem with an MSE implementation? Adding the duration of the ad or not adding it?

From an MSE implementation it looks the same as we apply an offset in both cases. However, I would prefer to test the VoD case, not adding the duration of the ad. Main argument here is that for live the `@presentationTimeOffset` in period 3 could be omitted when setting `@startnumber` to 1 and `@t` to 0.

**Comment by rbouqueau** [comment]
FYI there already is a version of the content with ToS and BBB as an ad: https://reference.dashif.org/dash.js/nightly/samples/dash-if-reference-player/index.html?mpd=https://dash.akamaized.net/dashif/ad-insertion-testcase1/batch5/real/a/ad-insertion-testcase1.mpd

**Comment by andyburras** [comment]
We need to use the mezzanine content for both the main and advert because we need the QR codes to be able to check requirements like all the frames of the advert are presented in the correct order. So I don't think reusing any DASH-IF content is an option.

**Comment by jpiesing** [comment]
> FYI there already is a version of the content with ToS and BBB as an ad: https://reference.dashif.org/dash.js/nightly/samples/dash-if-reference-player/index.html?mpd=https://dash.akamaized.net/dashif/ad-insertion-testcase1/batch5/real/a/ad-insertion-testcase1.mpd

Unfortunately that doesn't have the annotations that the observation framework relies on.

**Comment by rbouqueau** [comment]
Thanks both for the clarification. I can regenerate the same content with the mezzaning source whenever you feel this is ready to process. Let me know.

**Comment by jpiesing** [comment]
> Thanks both for the clarification. I can regenerate the same content with the mezzanine source whenever you feel this is ready to process. Let me know.

@louaybassbouss @rbouqueau Is there any reason not to go ahead with this now & take advantage of the time while the encryption issue is being investigated?

**Comment by louaybassbouss** [comment]
> > Thanks both for the clarification. I can regenerate the same content with the mezzanine source whenever you feel this is ready to process. Let me know.
> 
> @louaybassbouss @rbouqueau Is there any reason not to go ahead with this now & take advantage of the time while the encryption issue is being investigated?

@jpiesing from test implementation perspective we already tested the latest Multi-Period content created by @rbouqueau (this one https://dash.akamaized.net/dashif/ad-insertion-testcase1/batch5/real/a/ad-insertion-testcase1.mpd) and it is working fine (video plays until the end as expected). Once the Multi-Period content with mezzanine source is available, we will do final check that the new content is still working fine.

**Comment by jpiesing** [comment]
> @jpiesing from test implementation perspective we already tested the latest Multi-Period content created by @rbouqueau (this one https://dash.akamaized.net/dashif/ad-insertion-testcase1/batch5/real/a/ad-insertion-testcase1.mpd) and it is working fine (video plays until the end as expected). Once the Multi-Period content with mezzanine source is available, we will do final check that the new content is still working fine.

> Thanks both for the clarification. I can regenerate the same content with the mezzanine source whenever you feel this is ready to process. Let me know.

@rbouqueau  Please go ahead.

**Comment by nicholas-fr** [comment]
@rbouqueau I've just re-generated and uploaded the 30s bbb content so that the start/end frames have QR codes, in line with the latest mezzanine changes.
The new streams are here: https://dash.akamaized.net/WAVE/Mezzanine/under_review/2021-06-30/
720p/1080p, 25fps/30fps for now, for an initial test/review.

**Comment by FritzHeiden** [comment]
I added a test with multi period content (https://dash.akamaized.net/dashif/ad-insertion-testcase1/batch5/real/a/ad-insertion-testcase1.mpd) in case you want to test it.

**Comment by rbouqueau** [comment]
Thanks for assembling the pieces.

I re-executed the DASH-IF splice script from the 25 fps content ; note that it is a completely different script from the one executed for the existing CTA-Wave content.

I kept the same timings and durations as in the DASH-IF script. Let me know if you need something else:
https://reference.dashif.org/dash.js/nightly/samples/dash-if-reference-player/index.html?mpd=https://www.gpac-licensing.com/downloads/CTA-Wave/splice-testcase1/batch1/ad-insertion-testcase1.mpd

I plan to upload them to the Akamai server. Should we host them in the WAVE/vectors folder?

**Comment by jpiesing** [comment]
> Should we host them in the WAVE/vectors folder?

Yes please. In order to integrate it with all the rest of the work from Fraunhofer, I think it needs to be included in database.json as well.

@louaybassbouss @FritzHeiden Do you need anything else to add this to the CSV file & generate an HTML+JS+MPD combination that I can run?

**Comment by louaybassbouss** [comment]
> Thanks for assembling the pieces.
> 
> I re-executed the DASH-IF splice script from the 25 fps content ; note that it is a completely different script from the one executed for the existing CTA-Wave content.
> 
> I kept the same timings and durations as in the DASH-IF script. Let me know if you need something else:
> https://reference.dashif.org/dash.js/nightly/samples/dash-if-reference-player/index.html?mpd=https://www.gpac-licensing.com/downloads/CTA-Wave/splice-testcase1/batch1/ad-insertion-testcase1.mpd
> 
> I plan to upload them to the Akamai server. Should we host them in the WAVE/vectors folder?

@rbouqueau we couldn't test the [latest content](https://www.gpac-licensing.com/downloads/CTA-Wave/splice-testcase1/batch1/ad-insertion-testcase1.mpd) it seems that CORS header are not set. You can proceed as I mentioned in previous comment with uploading the content to WAVE/vectors folder and updating database.json file. Please ping us when this is done.

**Comment by rbouqueau** [comment]
Ok. I would like someone to validate the content itself. Is it ok?

https://reference.dashif.org/dash.js/nightly/samples/dash-if-reference-player/index.html?mpd=https://www.gpac-licensing.com/downloads/CTA-Wave/splice-testcase1/batch1/ad-insertion-testcase1.mpd

**Comment by louaybassbouss** [comment]
> Ok. I would like someone to validate the content itself. Is it ok?
> 
> https://reference.dashif.org/dash.js/nightly/samples/dash-if-reference-player/index.html?mpd=https://www.gpac-licensing.com/downloads/CTA-Wave/splice-testcase1/batch1/ad-insertion-testcase1.mpd

![image](https://user-images.githubusercontent.com/1138409/124089119-e1c72e80-da53-11eb-9c92-bb5bce47a438.png)
 but the MPD file is available

**Comment by jpiesing** [comment]
Please don't forget that test 8.13 is "Restricted Splicing of Encrypted Content". My interpretation is that this is the same as 8.8 but everything is encrypted. Transitions between encrypted & un-encrypted content and back again are covered by 8.14.

Obviously encryption will be blocked by https://github.com/cta-wave/dpctf-tests/issues/12 .

**Comment by nicholas-fr** [comment]
I played the content using VLC.
I was expecting main > ad > main.
Instead ad > main > ad mezzanine has been used. So I think the content used need to be switched.

More precisely I would expect to see:
Start frame of main > main frames > start frame of ad > ad frames > end frame of ad > more main frames > end frame of main

Regarding the durations, I would expect that the mezzanine for the 2nd period (inserted ad) needs to be of the exact duration of that period, so we see the start/end frames.
Based on the current mezzanine that would mean a total content duration of 90s (30s main + 30s ad + 30s main).
I see it's currently 9.6s+9.6s+9.6s.

**Comment by rbouqueau** [comment]
In case it can help, to bypass CORS restrictions you can find browser extensions.

@nicholas-fr Thanks for the clarifications, this is what I was expecting. FYI if we do 30s then the media durations won't be aligned with the period durations ; not an issue per se but that's a contraint that was not present in the original DASH-IF test.

**Comment by jpiesing** [comment]
> @nicholas-fr Thanks for the clarifications, this is what I was expecting. FYI if we do 30s then the media durations won't be aligned with the period durations ; not an issue per se but that's a contraint that was not present in the original DASH-IF test.

For the basic WAVE spicing test, we need the Period duration to be an exact multiple of the video & audio segment durations so there no gaps & no overlaps.
@nicholas-fr  That means we will need new mezzanine content :(

**Comment by haudiobe** [comment]
Colleagues, I need to chime in here. I just talked to Romain and afterwards to Daniel. I believe we are making a wrong assumption, namely to create a spliced content. The basic idea of all of the work is to 
1) use 2 content items
2) instruct how they are spliced in time

I attach a slide deck that provides the basics for this.
[Splice-Conditioning.pptx](https://github.com/cta-wave/dpctf-tests/files/6748477/Splice-Conditioning.pptx)

We need to update the description in the DPC spec, I admit.

I will try to join next week for the call to fix all of this.

Thanks

**Comment by jpiesing** [comment]
Thomas,


Can I ask why you believe this?


Jon


________________________________

**Comment by haudiobe** [comment]
I am ok either way. But I fear that we have to a lot of detailed work in front of us for every possible splice constraints and a new MPD needs to be created. The test stimulus then is a DASH client implementation rather than a description of the functionality. Even if we create these MPDs for splicing, we should do some scripting to say: create a splice. Manual creation will not scale.

So I am ok either way, but I want to avoid that Romain spends hours and hours on creating manual MPDs for splicing.

**Comment by louaybassbouss** [comment]
> I am ok either way. But I fear that we have to a lot of detailed work in front of us for every possible splice constraints and a new MPD needs to be created. The test stimulus then is a DASH client implementation rather than a description of the functionality. Even if we create these MPDs for splicing, we should do some scripting to say: create a splice. Manual creation will not scale.
> 
> So I am ok either way, but I want to avoid that Romain spends hours and hours on creating manual MPDs for splicing.

@haudiobe the current implementation of the common player accepts 2 MPDs one for video and one for audio as it was the requirement until now. Changing the test player to handle more MPDs for video is time consuming for us and we don't have the resources to do this change in the player.  In addition we need to make sure that other tests are not affected. From our perspective, we prefer to option with the Multi-Period as we agreed in previous test calls. Especially that we investigated time to implement Multi-Period to support this test case.

**Comment by haudiobe** [comment]
Well, multi-period is kind of possible. But it requires lots of work in the spec because we need to exactly describe the different cases and secondly, people will complain again why we do not do the same for HLS. And at the end the whole issue gets more complex even if you combine audio and video and so on. the MPD in the DPC spec where NEVER considered being playback instructions, they were always considered as content annotation. I am worried that we go into very different path and all of a sudden do a player test rather than a playback platform test.

**Comment by jpiesing** [comment]
@haudiobe What we need to test is that the right sequence of CMAF segments are fed into an MSE SourceBuffer in the right order (and perhaps at the right time) and play correctly. It *should* not matter how the test code is instructed what that right sequence of CMAF media segments is. @louaybassbouss believes multi-Period is the simplest way to define what this right sequence of media segments is.

**Comment by haudiobe** [comment]
Ok - just for once. I will not argue against this anymore as I do not have the time to convince differently. But I want to say it here and clearly (and not not having said it): I am personally convinced that the approach is wrong and will not scale in the future as we will result in many cases that we need to indirectly support by MPDs instead of playback instructions. I am not sure who will be responsible of writing all of these MPDs whenever we want to splice different codecs, different frame rates, encrypted cases and so on. It is also complex if you have audio and video combined. At the end someone will have to write an automation script that gets input of instructions, this will generate an MPD, and this MPD will then be used for playback tests (basically developing a DASH player again). Romain will not be able to do this as part of Qualcomm work.

**Comment by jpiesing** [comment]
> Ok - just for once. I will not argue against this anymore as I do not have the time to convince differently. But I want to say it here and clearly (and not not having said it): I am personally convinced that the approach is wrong and will not scale in the future as we will result in many cases that we need to indirectly support by MPDs instead of playback instructions. I am not sure who will be responsible of writing all of these MPDs whenever we want to splice different codecs, different frame rates, encrypted cases and so on. It is also complex if you have audio and video combined. At the end someone will have to write an automation script that gets input of instructions, this will generate an MPD, and this MPD will then be used for playback tests (basically developing a DASH player again). I am not sure why we need to take this intermediate step that causes lots of unnecessary work. Romain will not be able to do this as part of Qualcomm work.

@haudiobe Two genuine questions;
* What is the alternative? Separate MPD for the content to be inserted & define a time to switch from one to the other?
* How bad is the scalability issue really?
I just want to make sure that we take a path that improves things & which can be delivered.

**Comment by andyburras** [comment]
I recall Kilroy did do some initial analysis of the combinatorial scale before he retired. I've just had a look in the CSTF and DPCTF documents register but couldn't find any released document that gave actual figures, so it may just have been something he discussed? Bill Rose gave an initial overview of the scope in https://standards.cta.tech/apps/org/workgroup/cstf/download.php/22077/latest 

Kilroy did propose a "playlist" format to drive the test execution which could access chunk/segments from any content (in https://standards.cta.tech/apps/org/workgroup/dpctf/download.php/24093/latest ). Hence the content was kept as simple as possible and the MPDs only served to describe the content (not to drive the playback). I.e. the complexity was kept in the test framework code, rather than in the generated content.
Maybe eventually a subset of this approach, adding the info to tests.csv could be used to drive the splicing.
I appreciate this would be a bunch of extra work, so we may need to do an interim MPD-driven approach for the time being to get the intial batch of tests out there, with a view to enhancing the system in a future version.

**Comment by rbouqueau** [comment]
I'm starting to think that we should separate the MPD generation from the media file generation. That's possible with GPAC and it is probably desirable for clarity in this project. What do you think?

**Comment by jpiesing** [comment]
As far as I can see, what we are looking for is something like this;

1) Encode a version of the main content with segment lengths such that video & audio periodically align. This **should** just be another variant along with streams T1, T2, .... T34. There is already support for varying the segment length in those scripts so this should not be too time consuming. This can be made a standard feature of those scripts.
2) Split the single Period of the main content into 2 Periods at a time when video and audio segments are aligned. In an ideal world, this would be done when the content was initially generated.
3) Encode a version of the content to be spliced & generate the MPD
4) Modify the MPD from the main content to insert the period from the spliced content between the two pre-existing Periods & stitch-up anything needed to make it work.

**Comment by jpiesing** [comment]
Here is my attempt to understand what I think @haudiobe is proposing. Improvements are welcome.

The test has the following;
* Two pieces of content, main and spliced
* The test configuration defines a number K, which is the number of segments/chunks from the start of the main content to the splice point
* The test code takes segments/chunks starting from 1 going up to K and appends these to the MSE source buffer
* Once segment/chunk K has been appended
* The test code sets timestampOffset to account for the spliced content starting from time=0 
* The test code appends all the segments/chunks from the spliced content from beginning to end
* Once the last segment/chunk of the spliced content has been appended ...
* The test code sets timestampOffset to account for the starting time of segment/chunk K+1 from the main content
* The test code appends the segments/chunks of the main content starting with K+1 and going on to the end

This is basically a subset of what's in section 8.8 of the DPCTF spec & what is defined there could do this if the playout[] array was constructed carefully. It's just not remotely obvious.

**Comment by louaybassbouss** [comment]
> Here is my attempt to understand what I think @haudiobe is proposing. Improvements are welcome.
> 
> The test has the following;
> 
> * Two pieces of content, main and spliced. Each is a CMAF switching set described by a DASH MPD. For 8.8 & 8.13, each has one CMAF video track. For 9.4, each as one CMAF video track and one CMAF audio track.
> * The test configuration defines a number K, which is the number of segments/chunks from the start of the main content to the splice point (for 9.4, there would be different values of K for video & audio)
> * The test code takes segments/chunks starting from 1 going up to K and appends these to the MSE source buffer
> * Once segment/chunk K has been appended ...
> * The test code sets timestampOffset to account for the spliced content starting from time=0
> * The test code appends all the segments/chunks from the spliced content from beginning to end
> * Once the last segment/chunk of the spliced content has been appended ...
> * The test code sets timestampOffset to account for the starting time of segment/chunk K+1 from the main content
> * The test code appends the segments/chunks of the main content starting with K+1 and going on to the end
> 
> This is basically a subset of what's in section 8.8 of the DPCTF spec & what is defined there could do this if the playout[] array was constructed carefully. It's just not remotely obvious.
> 
> playout[i]: Provides the triple (Switching Set, CMAF track number, Fragment number) for every playout position i=1,…,N that is be played out.
> 
> For example, consider switching after 20 segments, the spliced content having 15 segments/chunks and the main content having 30 segments/chunks.
> 
> playout = {
> { main, 1, 1 },
> { main, 1, 2 },
> ....
> { main, 1, 20 },
> { spliced, 1, 1 },
> ...
> { spliced 1, 15 },
> { main, 1, 21 },
> ...
> { main, 1, 30 }
> }
> 

Thanks @jpiesing this approach LGTM. I will make a proposal in another GH issue to address the changes in the workflow (starting from tests.csv) in order to support this approach. 

> Looking again at the DPCTF spec, this includes appending new headers when this is needed. It does not include setting timestampOffset which I think is a bug?

The current implementation sets `timestampOffset`  but yes it is better to make it clear in the spec.

**Comment by jpiesing** [comment]
Just to be clear, my previous comment was really my thought process. IMHO the only things wrong with what's in the spec for 8.8 are 1) missing reference to setting timestampOffset and 2) it really needs an example of what the playout array somewhat like the one I gave. (Assuming this is what was intended).

**Comment by louaybassbouss** [comment]
> Just to be clear, my previous comment was really my thought process. 

yes I interpreted it in this way and this is a good start point for our meeting next week 

> IMHO the only things wrong with what's in the spec for 8.8 are 1) missing reference to setting timestampOffset 
and 2) it really needs an example of what the playout array somewhat like the one I gave. (Assuming this is what was intended).

Regarding `playout` array we need to make clear how to address the Switching Set. My proposal is: 

* Keep the structure of the [tests.csv](https://github.com/cta-wave/dpctf-tests/blob/master/tests.csv) as it is (4 columns) but we should allow to specify additional video and audio MPDs in additional lines (keep Test ID empty). Example:

<html><body>


Test ID | Video MPD | Audio MPD | Folder
-- | -- | -- | --
playback-over-wave-baseline-splice-constraints | path/to/video/main.mpd | path/to/audio/main.mpd | My Folder
  &nbsp; | path/to/video/spliced.mpd | path/to/audio/spliced.mpd |   

</body>
</html>

* The Switching Set in the playout array is just the index. in this example main has index `1` and spliced `2` (in the order how they appear in the csv file).

**Comment by jpiesing** [comment]
From the July 20th meeting.

- What additional resolutions do we need for the "ad"?
> - Beyond 720p and 1080p
> - Proposal: no.
- What additional frame rates do we need for the "ad"?
> - Beyond 25 and 30
> - Proposal: no – don't bother with 50 and 60
- Do we want to do fractional frame rates & what duration would be appropriate for the "ad"?
> - Proposal: 21.25 (and a number of other digits of precision)
> - New github issue - https://github.com/cta-wave/mezzanine/issues/40
- Do we want to create new (shorter) main content?
> - How much shorter?
> - Proposal: New 10s mezzanine streams for 1080 & 720 at 25, 30, 29.97

**Comment by louaybassbouss** [comment]
@jpiesing @yanj-github we are mostly done with the implementation of supporting multiple input Audio/Video MPDs (SwitchingSets) to replace the Multi-Period implementation (Multi-Period support is still there but not used in any test). The code is available in a separate branch [feature-multi-mpd](https://github.com/cta-wave/dpctf-tests/tree/feature-multi-mpd) for now, before we merge with the master, we need to clarify few things related to the interfaces with the OF mainly the structure of [/generated/tests.json](https://github.com/cta-wave/dpctf-tests/blob/feature-multi-mpd/generated/tests.json). 

### Current structure of one item in tests.json (structure does not support multiple MPDs):

```yaml
{
    "path": "avc_15_30_60-2021-07-07-online/sequential-track-playback__t1.html",
    "video": "https://dash.akamaized.net/WAVE/vectors/2021-07-07/avc_sets/15_30_60/t1/stream.mpd",
    "audio": "",
    "code": "sequential-track-playback.html",
    "config": {
        "cmaf_track_duration": "PT0H1M0.000S",
        "fragment_duration": "PT0H0M2.021S",
        "representations": {
            "1": {
                "period": 1,
                "duration": 60,
                "type": "video",
                "fragment_duration": 1.92
            }
        }
    }
}
```

### This is a proposal for the structure of each item in [/generated/tests.json](https://github.com/cta-wave/dpctf-tests/blob/feature-multi-mpd/generated/tests.json) with support of multiple MPDs/SwitchingSets:

```yaml
{
    "path": "avc_15_30_60-2021-07-07-online/sequential-track-playback__t1_t2.html",
    "video": [
        "https://dash.akamaized.net/WAVE/vectors/2021-07-07/avc_sets/15_30_60/t1/stream.mpd",
        "https://dash.akamaized.net/WAVE/vectors/2021-07-07/avc_sets/15_30_60/t2/stream.mpd"
    ],
    "audio": [,],
    "code": "sequential-track-playback.html",
    "switchingSets": {
        "video": [{
            "cmaf_track_duration": "PT0H1M0.000S",
            "fragment_duration": "PT0H0M1.920S",
            "representations": {
                "1": {
                    "period": 1,
                    "duration": 60,
                    "type": "video",
                    "fragment_duration": 1.92
                }
            }
        },{
            "cmaf_track_duration": "PT0H0M9.600S",
            "fragment_duration": "PT0H0M1.920S",
            "representations": {
                "1": {
                    "period": 1,
                    "duration": 9.6,
                    "type": "video",
                    "fragment_duration": 1.92
                }
            }
        }],
        "audio": [,]
    }
}
```

### Changes:

1. `video` and `audio` properties are arrays and not single values
2. the generated `path` name will consider the Stream ID of each MPD (in this example `__t1_t2.html` at the end of the filename).
3.  rename `config` in `switchingSets` (easier to map to spec texts)
    * `switchingSets` includes two elements `video` and `audio`. Each of them is an array including information extracted from the corresponding MPD. 

Below is an example of `playout` array (play first 5 segments from first switchingset, the first 5 segments from second switchingset and then the remaining  segments of the first switchingset):

```yaml
{ 
    "playout": [[1,1,1], [1,1,2], [1,1,3], [1,1,4], [1,1,5], [2,1,1], [2,1,2], [2,1,3], [2,1,4], [2,1,5], [1,1,6], [1,1,7], [1,1,8], [1,1,9], [1,1,10] ...]
}
```

**Comment by jpiesing** [comment]
@louaybassbouss I have no comments on the proposal.
@yanj-gitub what do you think?

**Comment by yanj-github** [comment]
Thanks @louaybassbouss  
The proposed configuration looks good to me for splicing test.
I can see the "config" changed to "switchingSets" and elements are changed to a list.
Are you going to change this to other configuration as well or only for splicing tests please?
What are the configuration look like for single mpd tests such as sequential-track-playback?

**Comment by louaybassbouss** [comment]
> Thanks @louaybassbouss
> The proposed configuration looks good to me for splicing test.
> I can see the "config" changed to "switchingSets" and elements are changed to a list.
> Are you going to change this to other configuration as well or only for splicing tests please?
> What are the configuration look like for single mpd tests such as sequential-track-playback?

Our proposal is to apply to all tests (). For other tests, the length of arrays will be `1`. Example:

```yaml
{
    "path": "avc_15_30_60-2021-07-07-online/sequential-track-playback__t1.html",
    "video": [
        "https://dash.akamaized.net/WAVE/vectors/2021-07-07/avc_sets/15_30_60/t1/stream.mpd",
    ],
    "audio": [],
    "code": "sequential-track-playback.html",
    "switchingSets": {
        "video": [{
            "cmaf_track_duration": "PT0H1M0.000S",
            "fragment_duration": "PT0H0M1.920S",
            "representations": {
                "1": {
                    "period": 1,
                    "duration": 60,
                    "type": "video",
                    "fragment_duration": 1.92
                }
            }
        }],
        "audio": []
    }
}
```

wdyt?

**Comment by yanj-github** [comment]
Thanks @louaybassbouss and @jpiesing 
I have following questions if you can help, please?
Q1 to Q3 are question about how we stop test configuration fault which will cause OF fault when it is configured in wrong way. OF will raise error or won’t be able to provide correct observation. It would be helpful either we document it clearly or throw error message when user intend to configure the tests in wrong way. Feel free to provide any comments and ideas.

Q4 and Q6 are question about the configuration file. 

Q1: In terms of number of mpds to configure for each test, how do we stop configuration fault? Will TR stop user from doing the following by giving error message or have clear documentation to user?
e.g: 2 mpds configured for non-splicing tests such as sequential test. 
       1 mpd configured for splicing tests
       3 mpd configured for splicing tests

Q2: OF will only work when playout is configured for switching set tests and splicing tests. Will TR report failure or when playout is missing?

Q3: The coratia and TOS mezzanine content has same ID. If user configure splice main and ad as same content ID such as coatia A1 and TOS A1 OF won’t be able to handle it due to detected recording has same content ID. Will TR reports error on this or we have clear documentation which state the test should not configured in this way?

Q4: Regarding to the playout for switching sets test, will it remain as it is now for switching set or it will be changed?

Q5: Can we have “frame rate” added in tests.json file please? OF can detect it from QR code but it is helpful for us to know frame rate to expect before OF detect the QR code.

Q6: Is there any limitation of playout for splicing test? Or it will be robust that any values are allowed please?

**Comment by FritzHeiden** [comment]
I implemented the proposed changes to the structure of the tests.json and regenerated it. You can find the new [tests.json](https://github.com/cta-wave/dpctf-tests/blob/feature-multi-mpd/generated/tests.json) in the [feature-multi-mpd](https://github.com/cta-wave/dpctf-tests/blob/feature-multi-mpd) branch. I've created a [branch](https://github.com/cta-wave/dpctf-deploy/tree/multi-mpd) in the deploy repository for testing it.

**Comment by yanj-github** [comment]
Hi @louaybassbouss and @FritzHeiden will you be able to add “frame rate” to the "tests.json" file please? It is currently a blocker for us. Thanks.

**Comment by FritzHeiden** [comment]
I updated the generate script to include the video frame rate. You can find it on the [feature-multi-mpd](https://github.com/cta-wave/dpctf-tests/tree/feature-multi-mpd) branch.

**Comment by louaybassbouss** [comment]
@yanj-github this issue is ready from our side the only missing step is to merge it with master once you finished testing.

**Comment by yanj-github** [comment]
This is tracked on https://github.com/cta-wave/dpctf-tests/issues/22 we can close this as this is duplicated.

**Comment by louaybassbouss** [comment]
> Thanks @louaybassbouss and @jpiesing
> I have following questions if you can help, please?
> Q1 to Q3 are question about how we stop test configuration fault which will cause OF fault when it is configured in wrong way. OF will raise error or won’t be able to provide correct observation. It would be helpful either we document it clearly or throw error message when user intend to configure the tests in wrong way. Feel free to provide any comments and ideas.
> 
> Q4 and Q6 are question about the configuration file.
> 
> Q1: In terms of number of mpds to configure for each test, how do we stop configuration fault? Will TR stop user from doing the following by giving error message or have clear documentation to user?
> e.g: 2 mpds configured for non-splicing tests such as sequential test.
> 1 mpd configured for splicing tests
> 3 mpd configured for splicing tests

For all tests that expect a single MPD such as sequential test only the first combination will be considered. All other MPDs are ignored. There is no error message.

In the current set of tests, only Splicing tests expect more than 1 MPD and you can control it via the `playout` parameter. MPDs that are not addressed in the `playout` parameter are ignored. If an MPD index is not defined, then we throw an error and this will be reported in the QR Code. 

> 
> Q2: OF will only work when playout is configured for switching set tests and splicing tests. Will TR report failure or when playout is missing?

By default, the first representation of the first MPD is played. We can end the test with an error if this is preferred. 

> 
> Q3: The coratia and TOS mezzanine content has same ID. If user configure splice main and ad as same content ID such as coatia A1 and TOS A1 OF won’t be able to handle it due to detected recording has same content ID. Will TR reports error on this or we have clear documentation which state the test should not configured in this way?

If mezzanine content with same ID is used in the CSV, we add an increment to the filename. But the preferred way is to use different folders (last column in the CSV). You can still distinguish between the different content since the MPD URLs will be different.

> 
> Q4: Regarding to the playout for switching sets test, will it remain as it is now for switching set or it will be changed?
> Updates from last meeting: We agreed to have 3 sets od value and OF will ignore the 1st value which is swithing set numbers.

I hope this is answered in last meeting

> 
> Q5: Can we have “frame rate” added in tests.json file please? OF can detect it from QR code but it is helpful for us to know frame rate to expect before OF detect the QR code. This is a blocker for us, if you can confirm this can be done it would be helpful.
> 

This is now added to the feature-multi-mpd

> Q6: Is there any limitation of playout for splicing test? Or it will be robust that any values are allowed please?

What limitations do you have in mind

**Comment by yanj-github** [comment]
Thanks @louaybassbouss 
Q1
> For all tests that expect a single MPD such as sequential test only the first combination will be considered. All other MPDs are ignored. There is no error message.
> 
> In the current set of tests, only Splicing tests expect more than 1 MPD and you can control it via the playout parameter. MPDs that are not addressed in the playout parameter are ignored. If an MPD index is not defined, then we throw an error and this will be reported in the QR Code.

This sounds good to me. From the branch I can see that multiple MPDs configured for switching test which confuses me. If multiple MPDs only expected for splicing tests can you kindly change it please?
https://github.com/cta-wave/dpctf-tests/blob/feature-multi-mpd/tests.csv

Q2
> By default, the first representation of the first MPD is played. We can end the test with an error if this is preferred.

When playout is missing for the test which it is expected, OF will give a result saying playout parameter is missing. User will then find out configuration missing after finished the observation. I would prefer to know playout is missing earlier if I were a user. 

Q3
> If mezzanine content with same ID is used in the CSV, we add an increment to the filename. But the preferred way is to use different folders (last column in the CSV). You can still distinguish between the different content since the MPD URLs will be different.

Thanks for this. But I think I have caused confusion here. I meant the content ID is the ID from the QR code. TOS and Croatia mezzanine have same content ID in its QR code. When user define to use same mpd for splicing tests or use mpds has content ID for splicing tests OF will reports the fault as it failed to detect the playback changes from one content to another. 

Q6
> Is there any limitation of playout for splicing test? Or it will be robust that any values are allowed please?

with "playout": [[x,y,z],.. ...]
x is 1 or 2: do you allow any other values?
y is 1: do you allow any other values within the switching set? 
z is it always in order and covers whole duration?

**Comment by yanj-github** [comment]
@louaybassbouss  and @FritzHeiden 
Current configuration settings:
```
  "avc_15_30_60-2021-07-07-online/playback-over-wave-baseline-splice-constraints__t1_t2.html": {
    "playout": [
      [1, 1, 1],  [1, 1, 2],  [1, 1, 3],  [1, 1, 4],  [1, 1, 5],
      [1, 1, 6],  [1, 1, 7],  [1, 1, 8],  [1, 1, 9],  [1, 1, 10],
      [1, 1, 11], [1, 1, 12], [1, 1, 13], [1, 1, 14], [1, 1, 15],
      [2, 1, 7],  [2, 1, 8],  [2, 1, 9],
      [2, 1, 10], [2, 1, 11], [2, 1, 12]
    ]
  },
```
Just plays back from start to end there is no splicing.

I have also tested the following playout:
```
  "avc_15_30_60-2021-07-07-online/playback-over-wave-baseline-splice-constraints__t1_t2.html": {
    "playout": [
      [1, 1, 1],  [1, 1, 2],  [1, 1, 3],  [1, 1, 4],  [1, 1, 5],
      [1, 1, 6],  [1, 1, 7],  [1, 1, 8],  [1, 1, 9],  [1, 1, 10],
      [1, 1, 11], [1, 1, 12], [1, 1, 13], [1, 1, 14], [1, 1, 15],
      [2, 1, 1],  [2, 1, 2],  [2, 1, 3],
      [2, 1, 4], [2, 1, 4], [2, 1, 6],
      [2, 1, 7],  [2, 1, 8],  [2, 1, 9],
      [2, 1, 10], [2, 1, 11], [2, 1, 12]
    ]
  },       
```
This does splice after 30 mins and it back to begining. This is as expected.
However the test timed out after [2, 1, 12]
and the result shows "timedout" with "undefined" from the error message. Can you have a look please?

Also t1 and t2 has same content ID which makes is unable for OF to detected the change in content. Switching set should be used so can we change main ss1 and While abosent in proper splicing content (big bucket bunny), can we use t21 please? t21 has content id "K1" different to what we have in ss1.

**Comment by louaybassbouss** [comment]
> @louaybassbouss and @FritzHeiden Current configuration settings:
> 
> ```
>   "avc_15_30_60-2021-07-07-online/playback-over-wave-baseline-splice-constraints__t1_t2.html": {
>     "playout": [
>       [1, 1, 1],  [1, 1, 2],  [1, 1, 3],  [1, 1, 4],  [1, 1, 5],
>       [1, 1, 6],  [1, 1, 7],  [1, 1, 8],  [1, 1, 9],  [1, 1, 10],
>       [1, 1, 11], [1, 1, 12], [1, 1, 13], [1, 1, 14], [1, 1, 15],
>       [2, 1, 7],  [2, 1, 8],  [2, 1, 9],
>       [2, 1, 10], [2, 1, 11], [2, 1, 12]
>     ]
>   },
> ```
> 
> Just plays back from start to end there is no splicing.
> 
> I have also tested the following playout:
> 
> ```
>   "avc_15_30_60-2021-07-07-online/playback-over-wave-baseline-splice-constraints__t1_t2.html": {
>     "playout": [
>       [1, 1, 1],  [1, 1, 2],  [1, 1, 3],  [1, 1, 4],  [1, 1, 5],
>       [1, 1, 6],  [1, 1, 7],  [1, 1, 8],  [1, 1, 9],  [1, 1, 10],
>       [1, 1, 11], [1, 1, 12], [1, 1, 13], [1, 1, 14], [1, 1, 15],
>       [2, 1, 1],  [2, 1, 2],  [2, 1, 3],
>       [2, 1, 4], [2, 1, 4], [2, 1, 6],
>       [2, 1, 7],  [2, 1, 8],  [2, 1, 9],
>       [2, 1, 10], [2, 1, 11], [2, 1, 12]
>     ]
>   },       
> ```
> 
> This does splice after 30 mins and it back to begining. This is as expected. However the test timed out after [2, 1, 12] and the result shows "timedout" with "undefined" from the error message. Can you have a look please?
> 
> Also t1 and t2 has same content ID which makes is unable for OF to detected the change in content. Switching set should be used so can we change main ss1 and While abosent in proper splicing content (big bucket bunny), can we use t21 please? t21 has content id "K1" different to what we have in ss1.

@yanj-github we will check. yes we can use `t21`

**Comment by yanj-github** [comment]
@jpiesing  Thanks very mich.
@louaybassbouss I am not really familar with the content downloading process between TR and the content pool.
Can you help with this issues please?
As Jon mentioned I think we should use splice_ad_bbb if avaliable.

**Comment by jpiesing** [comment]
> As Jon mentioned we should use splice_ad_bbb if avaliable.

If splice_ad_bbb_* don't exist in encoded forms then someone needs to open an issue in the test content generation project asking romain to encode them.

**Comment by yanj-github** [comment]
https://github.com/cta-wave/Test-Content/issues/20

**Comment by jpiesing** [comment]
Now we have https://github.com/cta-wave/dpctf-tests/blob/stable/generated/avc_12.5_25_50-local/playback-over-wave-baseline-splice-constraints__splice_main_splice_ad.html, can we close this issue?

**Comment by louaybassbouss** [comment]
> Now we have https://github.com/cta-wave/dpctf-tests/blob/stable/generated/avc_12.5_25_50-local/playback-over-wave-baseline-splice-constraints__splice_main_splice_ad.html, can we close this issue?

Yes closing
